// @1claw/sdk/dist/core/client.js
import { HttpClient } from "./http";
import { VaultResource } from "../resources/vault";
import { SecretsResource } from "../resources/secrets";
import { AccessResource } from "../resources/access";
import { AgentsResource } from "../resources/agents";
import { SharingResource } from "../resources/sharing";
import { ApprovalsResource } from "../resources/approvals";
import { BillingResource } from "../resources/billing";
import { AuditResource } from "../resources/audit";
import { OrgResource } from "../resources/org";
import { AuthResource } from "../resources/auth";
import { ApiKeysResource } from "../resources/api-keys";
import { ChainsResource } from "../resources/chains";
import { X402Resource } from "../resources/x402";
import { TreasuryResource } from "../resources/treasury";
/**
 * The main 1Claw SDK client. All API resources are exposed as
 * namespaced properties for discoverability and tree-shaking.
 *
 * @example
 * ```ts
 * import { createClient } from "@1claw/sdk";
 *
 * const client = createClient({
 *   baseUrl: "https://api.1claw.xyz",
 *   apiKey: "ocv_...",
 * });
 *
 * // Authenticate (exchanges API key for a JWT)
 * await client.auth.apiKeyToken({ api_key: "ocv_..." });
 *
 * // Use resources
 * const vaults = await client.vault.list();
 * const secret = await client.secrets.get(vaultId, "my-key");
 * ```
 */
export class OneclawClient {
    constructor(config) {
        this.http = new HttpClient(config);
        if (config.apiKey && !config.token && !config.agentId) {
            // ocv_ keys are agent keys — HttpClient handles token exchange.
            // 1ck_ keys are user keys — need a one-time token exchange here.
            if (!config.apiKey.startsWith("ocv_")) {
                this.autoAuthenticateUserKey(config);
            }
        }
        this.vault = new VaultResource(this.http);
        this.secrets = new SecretsResource(this.http);
        this.access = new AccessResource(this.http);
        this.agents = new AgentsResource(this.http);
        this.sharing = new SharingResource(this.http);
        this.approvals = new ApprovalsResource(this.http);
        this.billing = new BillingResource(this.http);
        this.audit = new AuditResource(this.http);
        this.org = new OrgResource(this.http);
        this.auth = new AuthResource(this.http);
        this.apiKeys = new ApiKeysResource(this.http);
        this.chains = new ChainsResource(this.http);
        this.x402 = new X402Resource(this.http, config.x402Signer);
        this.treasury = new TreasuryResource(this.http);
    }
    autoAuthenticateUserKey(config) {
        const authPromise = this.http
            .request("POST", "/v1/auth/api-key-token", {
            body: { api_key: config.apiKey },
        })
            .then((res) => {
            if (res.data?.access_token)
                this.http.setToken(res.data.access_token);
        });
        authPromise.catch(() => {
            /* auth failure will surface on the next request */
        });
    }
}
/**
 * Factory function to create a new 1Claw SDK client.
 *
 * @example
 * ```ts
 * // User with API key
 * const client = createClient({
 *   baseUrl: "https://api.1claw.xyz",
 *   apiKey: "ocv_...",
 * });
 *
 * // Agent with API key
 * const agent = createClient({
 *   baseUrl: "https://api.1claw.xyz",
 *   apiKey: "ocv_...",
 *   agentId: "agent-uuid",
 * });
 *
 * // Pre-authenticated with JWT
 * const authed = createClient({
 *   baseUrl: "https://api.1claw.xyz",
 *   token: "eyJ...",
 * });
 * ```
 */
export function createClient(config) {
    return new OneclawClient(config);
}
