// @1claw/sdk/dist/core/errors.js
/**
 * Base error class for all 1Claw SDK errors.
 * Includes the HTTP status code and a machine-readable error type.
 */
export class OneclawError extends Error {
    constructor(message, status, type, detail) {
        super(message);
        this.name = "OneclawError";
        this.status = status;
        this.type = type;
        this.detail = detail;
    }
}
/** Thrown on 401 Unauthorized or 403 Forbidden responses. */
export class AuthError extends OneclawError {
    constructor(message, status = 401) {
        super(message, status, "auth_error");
        this.name = "AuthError";
    }
}
/**
 * Thrown when a resource limit is exceeded (403 with type "resource_limit_exceeded").
 * The `detail` message includes the current count, limit, and tier.
 */
export class ResourceLimitExceededError extends OneclawError {
    constructor(message) {
        super(message, 403, "resource_limit_exceeded");
        this.name = "ResourceLimitExceededError";
    }
}
/**
 * Thrown on 402 Payment Required responses.
 * Contains the full `PaymentRequirement` so callers can inspect
 * price, network, and payment address before deciding to pay.
 */
export class PaymentRequiredError extends OneclawError {
    constructor(message, paymentRequirement) {
        super(message, 402, "payment_required");
        this.name = "PaymentRequiredError";
        this.paymentRequirement = paymentRequirement;
    }
}
/**
 * Thrown when a secret is gated behind human approval.
 * The `approvalRequestId` can be used to poll for approval status.
 */
export class ApprovalRequiredError extends OneclawError {
    constructor(approvalRequestId, message) {
        super(message ?? "Human approval is required to access this secret", 403, "approval_required");
        this.name = "ApprovalRequiredError";
        this.approvalRequestId = approvalRequestId;
    }
}
/** Thrown on 404 Not Found responses. */
export class NotFoundError extends OneclawError {
    constructor(message = "Resource not found") {
        super(message, 404, "not_found");
        this.name = "NotFoundError";
    }
}
/** Thrown on 429 Too Many Requests responses. Includes retry timing when available. */
export class RateLimitError extends OneclawError {
    constructor(message = "Rate limit exceeded", retryAfterMs) {
        super(message, 429, "rate_limit");
        this.name = "RateLimitError";
        this.retryAfterMs = retryAfterMs;
    }
}
/** Thrown on 400 Bad Request responses for validation failures. */
export class ValidationError extends OneclawError {
    constructor(message, fields) {
        super(message, 400, "validation_error");
        this.name = "ValidationError";
        this.fields = fields;
    }
}
/** Thrown on 500+ server-side errors. */
export class ServerError extends OneclawError {
    constructor(message = "Internal server error", status = 500) {
        super(message, status, "server_error");
        this.name = "ServerError";
    }
}
/**
 * Parse an HTTP response into the appropriate typed error.
 * Falls back to a generic `OneclawError` for unrecognised status codes.
 */
export async function errorFromResponse(res) {
    let body = {};
    try {
        body = (await res.json());
    }
    catch {
        /* response may not be JSON */
    }
    const message = body.detail ??
        body.message ??
        body.error ??
        `HTTP ${res.status}`;
    switch (res.status) {
        case 400:
            return new ValidationError(message, body.fields);
        case 401:
            return new AuthError(message, 401);
        case 403: {
            const errorType = body.type;
            if (errorType === "resource_limit_exceeded") {
                return new ResourceLimitExceededError(message);
            }
            if (errorType === "approval_required") {
                const approvalId = body.approval_request_id ?? "";
                return new ApprovalRequiredError(approvalId, message);
            }
            return new AuthError(message, 403);
        }
        case 402: {
            const requirement = body;
            return new PaymentRequiredError(message, requirement);
        }
        case 404:
            return new NotFoundError(message);
        case 429: {
            const retryHeader = res.headers.get("Retry-After");
            const retryAfterMs = retryHeader
                ? parseInt(retryHeader, 10) * 1000
                : undefined;
            return new RateLimitError(message, retryAfterMs);
        }
        default:
            if (res.status >= 500)
                return new ServerError(message, res.status);
            return new OneclawError(message, res.status, "unknown");
    }
}
