// @1claw/sdk/dist/core/http.js
import { errorFromResponse, PaymentRequiredError } from "./errors";
/**
 * Internal HTTP transport used by every resource module.
 * Handles authentication headers, 402 auto-pay, and error mapping.
 */
export class HttpClient {
    constructor(config) {
        this.tokenExpiresAt = 0;
        this.baseUrl = config.baseUrl.replace(/\/$/, "");
        this.token = config.token;
        this.signer = config.x402Signer;
        this.maxAutoPayUsd = config.maxAutoPayUsd ?? 0;
        const isAgentKey = config.apiKey?.startsWith("ocv_") || !!config.agentId;
        if (config.apiKey && isAgentKey) {
            this.agentCredentials = {
                agentId: config.agentId,
                apiKey: config.apiKey,
            };
        }
    }
    /** Agent ID resolved from the token exchange (for key-only auth). */
    get resolvedAgentId() {
        return this._resolvedAgentId ?? this.agentCredentials?.agentId;
    }
    /** Replace the current Bearer token (called by auth methods). */
    setToken(token) {
        this.token = token;
        this.tokenExpiresAt = HttpClient.decodeExpiry(token);
    }
    getToken() {
        return this.token;
    }
    getBaseUrl() {
        return this.baseUrl;
    }
    static decodeExpiry(jwt) {
        try {
            const parts = jwt.split(".");
            if (parts.length !== 3)
                return 0;
            const payload = JSON.parse(atob(parts[1].replace(/-/g, "+").replace(/_/g, "/")));
            return typeof payload.exp === "number" ? payload.exp * 1000 : 0;
        }
        catch {
            return 0;
        }
    }
    async ensureToken() {
        if (!this.agentCredentials)
            return;
        if (this.token &&
            Date.now() < this.tokenExpiresAt - HttpClient.REFRESH_BUFFER_MS)
            return;
        if (this.refreshPromise) {
            await this.refreshPromise;
            return;
        }
        this.refreshPromise = (async () => {
            const body = {
                api_key: this.agentCredentials.apiKey,
            };
            if (this.agentCredentials.agentId) {
                body.agent_id = this.agentCredentials.agentId;
            }
            const res = await fetch(`${this.baseUrl}/v1/auth/agent-token`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify(body),
            });
            if (!res.ok) {
                throw new Error(`Agent token refresh failed: HTTP ${res.status}`);
            }
            const data = (await res.json());
            this.token = data.access_token;
            this.tokenExpiresAt = Date.now() + data.expires_in * 1000;
            if (data.agent_id) {
                this._resolvedAgentId = data.agent_id;
                if (this.agentCredentials && !this.agentCredentials.agentId) {
                    this.agentCredentials.agentId = data.agent_id;
                }
            }
        })().finally(() => {
            this.refreshPromise = undefined;
        });
        await this.refreshPromise;
    }
    /**
     * Perform a typed request against the API and return an envelope.
     * Automatically retries once with an x402 payment header when
     * a signer is configured and the price is within limits.
     */
    async request(method, path, options = {}) {
        if (!options.skipAuth) {
            await this.ensureToken();
        }
        const url = this.buildUrl(path, options.query);
        const headers = {
            "Content-Type": "application/json",
            ...options.headers,
        };
        if (!options.skipAuth && this.token) {
            headers["Authorization"] = `Bearer ${this.token}`;
        }
        const init = { method, headers };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        let res = await fetch(url, init);
        if (res.status === 402 && this.signer) {
            res = await this.handlePayment(res, url, init);
        }
        if (!res.ok) {
            const err = await errorFromResponse(res);
            return {
                data: null,
                error: {
                    type: err.type,
                    message: err.message,
                    detail: err.detail,
                },
                meta: { status: res.status },
            };
        }
        if (res.status === 204) {
            return {
                data: null,
                error: null,
                meta: { status: 204 },
            };
        }
        const data = (await res.json());
        return { data, error: null, meta: { status: res.status } };
    }
    /**
     * Same as `request` but throws on error instead of returning an envelope.
     * Convenient for callers that prefer exceptions.
     */
    async requestOrThrow(method, path, options = {}) {
        await this.ensureToken();
        const url = this.buildUrl(path, options.query);
        const headers = {
            "Content-Type": "application/json",
            ...options.headers,
        };
        if (this.token) {
            headers["Authorization"] = `Bearer ${this.token}`;
        }
        const init = { method, headers };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        let res = await fetch(url, init);
        if (res.status === 402 && this.signer) {
            res = await this.handlePayment(res, url, init);
        }
        if (!res.ok) {
            throw await errorFromResponse(res);
        }
        if (res.status === 204) {
            return undefined;
        }
        return (await res.json());
    }
    // -----------------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------------
    buildUrl(path, query) {
        const url = new URL(`${this.baseUrl}${path}`);
        if (query) {
            for (const [k, v] of Object.entries(query)) {
                if (v !== undefined)
                    url.searchParams.set(k, String(v));
            }
        }
        return url.toString();
    }
    /**
     * Attempt to sign an x402 payment and retry the original request
     * with the `X-PAYMENT` header attached.
     */
    async handlePayment(originalRes, url, init) {
        let requirement;
        try {
            requirement = (await originalRes.json());
        }
        catch {
            return originalRes;
        }
        const accept = requirement.accepts?.[0];
        if (!accept || !this.signer)
            return originalRes;
        // Prefer maxAmountRequired (atomic); fall back to price (USD) for older servers
        const priceUsd = accept.price != null
            ? parseFloat(accept.price)
            : parseFloat(accept.maxAmountRequired) / 1000000; // USDC 6 decimals -> USD
        if (Number.isNaN(priceUsd)) {
            return originalRes;
        }
        if (this.maxAutoPayUsd > 0 && priceUsd > this.maxAutoPayUsd) {
            throw new PaymentRequiredError(`Payment of $${priceUsd.toFixed(4)} exceeds auto-pay limit of $${this.maxAutoPayUsd}`, requirement);
        }
        if (this.maxAutoPayUsd === 0) {
            throw new PaymentRequiredError(`Payment of $${priceUsd.toFixed(4)} required. Enable auto-pay via maxAutoPayUsd config.`, requirement);
        }
        const signature = await this.signer.signPayment(accept);
        const paymentPayload = {
            x402Version: requirement.x402Version,
            scheme: accept.scheme,
            network: accept.network,
            payload: signature,
        };
        const retryHeaders = {
            ...init.headers,
            "X-PAYMENT": JSON.stringify(paymentPayload),
        };
        return fetch(url, { ...init, headers: retryHeaders });
    }
}
HttpClient.REFRESH_BUFFER_MS = 60000;
