// @1claw/sdk/dist/resources/access.js
/**
 * Access resource — manage vault access policies (grants) for
 * humans and agents.
 */
export class AccessResource {
    constructor(http) {
        this.http = http;
    }
    /**
     * Grant a human user access to a vault.
     * @param permissions - e.g. ["read"], ["read", "write"]
     */
    async grantHuman(vaultId, userId, permissions, options = {}) {
        const body = {
            secret_path_pattern: options.secretPathPattern ?? "**",
            principal_type: "user",
            principal_id: userId,
            permissions,
            conditions: options.conditions,
            expires_at: options.expires_at,
        };
        return this.http.request("POST", `/v1/vaults/${vaultId}/policies`, { body });
    }
    /**
     * Grant an agent access to a vault.
     * @param permissions - e.g. ["read"], ["read", "write"]
     */
    async grantAgent(vaultId, agentId, permissions, options = {}) {
        const body = {
            secret_path_pattern: options.secretPathPattern ?? "**",
            principal_type: "agent",
            principal_id: agentId,
            permissions,
            conditions: options.conditions,
            expires_at: options.expires_at,
        };
        return this.http.request("POST", `/v1/vaults/${vaultId}/policies`, { body });
    }
    /** Update an existing policy's permissions and/or conditions. */
    async update(vaultId, policyId, update) {
        return this.http.request("PUT", `/v1/vaults/${vaultId}/policies/${policyId}`, { body: update });
    }
    /** Revoke a specific access policy by its ID. */
    async revoke(vaultId, policyId) {
        return this.http.request("DELETE", `/v1/vaults/${vaultId}/policies/${policyId}`);
    }
    /** List all access grants (policies) on a vault. */
    async listGrants(vaultId) {
        return this.http.request("GET", `/v1/vaults/${vaultId}/policies`);
    }
}
