// @1claw/sdk/dist/resources/agents.js
/**
 * Agents resource — register, manage, and rotate keys for AI agents
 * that interact with the vault programmatically.
 */
export class AgentsResource {
    constructor(http) {
        this.http = http;
    }
    /**
     * Register a new agent. Returns the agent record and a one-time API key.
     * Store the API key securely — it cannot be retrieved again.
     */
    async create(options) {
        return this.http.request("POST", "/v1/agents", {
            body: options,
        });
    }
    /**
     * Self-enroll an agent (public, no auth required).
     * Credentials are emailed to the human — they are NOT returned in the response.
     */
    async enroll(options) {
        return this.http.request("POST", "/v1/agents/enroll", { body: options });
    }
    /**
     * Static helper to self-enroll without an existing client instance.
     * Useful when the agent has no credentials yet.
     */
    static async enroll(baseUrl, options) {
        const res = await fetch(`${baseUrl}/v1/agents/enroll`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(options),
        });
        if (!res.ok) {
            const body = await res.json().catch(() => ({}));
            throw new Error(body.detail || body.message || `Enroll failed (${res.status})`);
        }
        return res.json();
    }
    /** Fetch the calling agent's own profile (includes `created_by`). */
    async getSelf() {
        return this.http.request("GET", "/v1/agents/me");
    }
    /** Fetch a single agent by ID. */
    async get(agentId) {
        return this.http.request("GET", `/v1/agents/${agentId}`);
    }
    /** List all agents in the current organization. */
    async list() {
        return this.http.request("GET", "/v1/agents");
    }
    /** Update agent name, scopes, active status, expiry, or Intents API setting. */
    async update(agentId, update) {
        return this.http.request("PATCH", `/v1/agents/${agentId}`, { body: update });
    }
    /** Delete an agent permanently. */
    async delete(agentId) {
        return this.http.request("DELETE", `/v1/agents/${agentId}`);
    }
    /**
     * Rotate an agent's API key. Returns the new key — store it securely.
     * The old key is immediately invalidated.
     */
    async rotateKey(agentId) {
        return this.http.request("POST", `/v1/agents/${agentId}/rotate-key`);
    }
    // ── Intents API ──────────────────────────────────────────────────
    /**
     * Submit a transaction intent to be signed by the Intents API.
     * The agent must have `intents_api_enabled: true` and a valid
     * signing key stored in an accessible vault.
     *
     * Automatically generates an Idempotency-Key header for replay
     * protection. Pass `idempotencyKey` to override with your own.
     */
    async submitTransaction(agentId, tx, options) {
        const key = options?.idempotencyKey ?? crypto.randomUUID();
        return this.http.request("POST", `/v1/agents/${agentId}/transactions`, { body: tx, headers: { "Idempotency-Key": key } });
    }
    /**
     * Fetch a single transaction by ID.
     * By default the API omits `signed_tx`; pass `includeSignedTx: true` to include it.
     */
    async getTransaction(agentId, txId, options) {
        const qs = options?.includeSignedTx === true
            ? "?include_signed_tx=true"
            : "";
        return this.http.request("GET", `/v1/agents/${agentId}/transactions/${txId}${qs}`);
    }
    /**
     * List recent transactions for an agent.
     * By default the API omits `signed_tx`; pass `includeSignedTx: true` to include it.
     */
    async listTransactions(agentId, options) {
        const qs = options?.includeSignedTx === true
            ? "?include_signed_tx=true"
            : "";
        return this.http.request("GET", `/v1/agents/${agentId}/transactions${qs}`);
    }
    /**
     * Sign a transaction without broadcasting. The signed_tx hex is returned
     * so the caller can submit to their own RPC endpoint.
     * All agent guardrails (allowlists, value caps, daily limits) are enforced.
     */
    async signTransaction(agentId, tx) {
        return this.http.request("POST", `/v1/agents/${agentId}/transactions/sign`, { body: tx });
    }
    // ── Transaction Simulation ─────────────────────────────────────────
    /**
     * Simulate a transaction via Tenderly without signing or broadcasting.
     * Returns balance changes, gas estimates, and success/revert status.
     */
    async simulateTransaction(agentId, tx) {
        return this.http.request("POST", `/v1/agents/${agentId}/transactions/simulate`, { body: tx });
    }
    /**
     * Simulate a bundle of transactions sequentially (e.g. approve + swap).
     */
    async simulateBundle(agentId, bundle) {
        return this.http.request("POST", `/v1/agents/${agentId}/transactions/simulate-bundle`, { body: bundle });
    }
}
