// @1claw/sdk/dist/resources/billing.js
/**
 * Billing resource — usage summaries, history, and LLM token billing.
 */
export class BillingResource {
    constructor(http) {
        this.http = http;
    }
    /** Get the current month's usage summary (free tier remaining, costs). */
    async usage() {
        return this.http.request("GET", "/v1/billing/usage");
    }
    /**
     * Get per-request usage history.
     * @param limit Maximum number of events to return (1–200, default 50).
     */
    async history(limit) {
        return this.http.request("GET", "/v1/billing/history", {
            query: limit !== undefined ? { limit } : undefined,
        });
    }
    // ── LLM Token Billing ──────────────────────────────────────────
    /** Get LLM token billing status for the current org. */
    async llmTokenBilling() {
        return this.http.request("GET", "/v1/billing/llm-token-billing");
    }
    /**
     * Subscribe to LLM token billing. Returns a Stripe Checkout URL.
     * Redirect the user to complete the subscription.
     */
    async subscribeLlmTokenBilling() {
        return this.http.request("POST", "/v1/billing/llm-token-billing/subscribe");
    }
    /** Disable LLM token billing for the current org. */
    async disableLlmTokenBilling() {
        return this.http.request("POST", "/v1/billing/llm-token-billing/disable");
    }
}
