// @1claw/sdk/dist/resources/chains.js
/**
 * Chains resource — list supported blockchains and manage chain configuration.
 *
 * Public endpoints let any authenticated user list enabled chains.
 * Admin endpoints allow adding, updating (e.g. setting an RPC URL), and removing chains.
 */
export class ChainsResource {
    constructor(http) {
        this.http = http;
    }
    /** List all enabled chains. */
    async list() {
        return this.http.request("GET", "/v1/chains");
    }
    /** Get a chain by name (e.g. "base") or numeric chain ID (e.g. "8453"). */
    async get(identifier) {
        return this.http.request("GET", `/v1/chains/${identifier}`);
    }
    // ── Admin ──────────────────────────────────────────────────────
    /** List all chains including disabled ones (admin). */
    async adminList() {
        return this.http.request("GET", "/v1/admin/chains");
    }
    /** Add a new chain to the registry (admin). */
    async create(chain) {
        return this.http.request("POST", "/v1/admin/chains", {
            body: chain,
        });
    }
    /** Update chain configuration, e.g. set an RPC URL (admin). */
    async update(chainId, update) {
        return this.http.request("PUT", `/v1/admin/chains/${chainId}`, { body: update });
    }
    /** Remove a chain from the registry (admin). */
    async delete(chainId) {
        return this.http.request("DELETE", `/v1/admin/chains/${chainId}`);
    }
}
