// @1claw/sdk/dist/resources/secrets.js
/**
 * Secrets resource — store, retrieve, list, rotate, and delete secrets
 * within a vault.
 */
export class SecretsResource {
    constructor(http) {
        this.http = http;
    }
    /**
     * Store or update a secret at the given path inside a vault.
     * Returns the secret metadata (without the plaintext value).
     */
    async set(vaultId, key, value, options = {}) {
        const body = {
            type: options.type ?? "generic",
            value,
            metadata: options.metadata,
            expires_at: options.expires_at,
            rotation_policy: options.rotation_policy,
            max_access_count: options.max_access_count,
        };
        return this.http.request("PUT", `/v1/vaults/${vaultId}/secrets/${key}`, { body });
    }
    /**
     * Retrieve a decrypted secret value.
     * May return a `PaymentRequiredError` (402) or `ApprovalRequiredError`
     * depending on access policies.
     */
    async get(vaultId, key, _options) {
        return this.http.request("GET", `/v1/vaults/${vaultId}/secrets/${key}`);
    }
    /** Delete a secret from a vault. */
    async delete(vaultId, key) {
        return this.http.request("DELETE", `/v1/vaults/${vaultId}/secrets/${key}`);
    }
    /** List secret keys (metadata only, no plaintext values). */
    async list(vaultId, prefix) {
        return this.http.request("GET", `/v1/vaults/${vaultId}/secrets`, { query: prefix ? { prefix } : undefined });
    }
    /**
     * Rotate a secret by writing a new value at the same path.
     * This increments the version and overwrites the previous value.
     */
    async rotate(vaultId, key, newValue, options = {}) {
        return this.set(vaultId, key, newValue, options);
    }
}
