// @1claw/sdk/dist/resources/treasury.js
/**
 * Treasury (Safe multisig) management — create, update, delete treasuries;
 * signers; agent access requests (approve/deny).
 */
export class TreasuryResource {
    constructor(http) {
        this.http = http;
    }
    async create(body) {
        return this.http.request("POST", "/v1/treasury", {
            body,
        });
    }
    async list() {
        return this.http.request("GET", "/v1/treasury");
    }
    async get(treasuryId) {
        return this.http.request("GET", `/v1/treasury/${treasuryId}`);
    }
    async update(treasuryId, body) {
        return this.http.request("PATCH", `/v1/treasury/${treasuryId}`, { body });
    }
    async delete(treasuryId) {
        return this.http.request("DELETE", `/v1/treasury/${treasuryId}`);
    }
    async addSigner(treasuryId, body) {
        return this.http.request("POST", `/v1/treasury/${treasuryId}/signers`, { body });
    }
    async removeSigner(treasuryId, signerId) {
        return this.http.request("DELETE", `/v1/treasury/${treasuryId}/signers/${signerId}`);
    }
    async requestAccess(treasuryId, body = {}) {
        return this.http.request("POST", `/v1/treasury/${treasuryId}/access-requests`, { body });
    }
    async listAccessRequests(treasuryId) {
        return this.http.request("GET", `/v1/treasury/${treasuryId}/access-requests`);
    }
    async approveAccess(treasuryId, requestId) {
        return this.http.request("POST", `/v1/treasury/${treasuryId}/access-requests/${requestId}/approve`);
    }
    async denyAccess(treasuryId, requestId) {
        return this.http.request("POST", `/v1/treasury/${treasuryId}/access-requests/${requestId}/deny`);
    }
}
