// @1claw/sdk/dist/resources/vault.js
/**
 * Vault resource — create, list, get, and delete encrypted vaults.
 */
export class VaultResource {
    constructor(http) {
        this.http = http;
    }
    /** Create a new vault. */
    async create(options) {
        return this.http.request("POST", "/v1/vaults", {
            body: {
                name: options.name,
                description: options.description ?? "",
            },
        });
    }
    /** Fetch a single vault by ID. */
    async get(vaultId) {
        return this.http.request("GET", `/v1/vaults/${vaultId}`);
    }
    /** List all vaults visible to the authenticated identity. */
    async list() {
        return this.http.request("GET", "/v1/vaults");
    }
    /** Permanently delete a vault and all its secrets. */
    async delete(vaultId) {
        return this.http.request("DELETE", `/v1/vaults/${vaultId}`);
    }
    /** Enable CMEK on a vault (Business/Enterprise only). */
    async enableCmek(vaultId, request) {
        return this.http.request("POST", `/v1/vaults/${vaultId}/cmek`, { body: request });
    }
    /** Disable CMEK on a vault. Secrets remain encrypted but no new CMEK layer is applied. */
    async disableCmek(vaultId) {
        return this.http.request("DELETE", `/v1/vaults/${vaultId}/cmek`);
    }
    /**
     * Start a server-assisted CMEK key rotation job.
     * Both keys are sent in headers (transit over TLS only) and are not logged.
     */
    async rotateCmek(vaultId, oldKeyBase64, newKeyBase64, newFingerprint) {
        return this.http.request("POST", `/v1/vaults/${vaultId}/cmek-rotate`, {
            body: { new_fingerprint: newFingerprint },
            headers: {
                "x-cmek-old-key": oldKeyBase64,
                "x-cmek-new-key": newKeyBase64,
            },
        });
    }
    /** Poll the status of a CMEK rotation job. */
    async getRotationJobStatus(vaultId, jobId) {
        return this.http.request("GET", `/v1/vaults/${vaultId}/cmek-rotate/${jobId}`);
    }
}
